File: patterns/tool-use-external-api/openai-like/javascript/crypto-price.ts
import 'dotenv/config';
import OpenAI from 'openai';

// Define the function interface
async function getCryptoPrice(coinId: string, currency: string = 'usd') {
    const url = `https://api.coingecko.com/api/v3/simple/price?ids=${coinId}&vs_currencies=${currency}`;
    try {
        const response = await fetch(url);
        const data = await response.json();
        const price = data[coinId] ? data[coinId][currency] : null;
        return JSON.stringify({ coin: coinId, price: price, currency: currency });
    } catch (error) {
        return JSON.stringify({ error: String(error) });
    }
}

async function main() {
    const apiKey = process.env.OPENAI_API_KEY;
    if (!apiKey) {
        console.error('Please set OPENAI_API_KEY.');
        return;
    }

    const client = new OpenAI({ apiKey });

    const tools: OpenAI.Chat.Completions.ChatCompletionTool[] = [
        {
            type: 'function',
            function: {
                name: 'get_crypto_price',
                description: 'Get the current price of a cryptocurrency',
                parameters: {
                    type: 'object',
                    properties: {
                        coin_id: {
                            type: 'string',
                            description: 'The ID of the coin on CoinGecko (e.g., bitcoin, ethereum)',
                        },
                        currency: {
                            type: 'string',
                            description: 'The target currency (e.g., usd, eur)',
                        },
                    },
                    required: ['coin_id'],
                },
            },
        },
    ];

    const messages: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = [
        { role: 'user', content: 'What is the price of Ethereum right now in USD?' },
    ];

    console.log(`User Question: ${messages[0].content}`);
    console.log('Sending request to OpenAI with tools...');

    const response = await client.chat.completions.create({
        model: 'gpt-3.5-turbo',
        messages: messages,
        tools: tools,
        tool_choice: 'auto',
    });

    const responseMessage = response.choices[0].message;
    const toolCalls = responseMessage.tool_calls;

    if (toolCalls) {
        console.log(`Model decided to call ${toolCalls.length} tool(s).`);
        messages.push(responseMessage);

        for (const toolCall of toolCalls) {
            const functionName = toolCall.function.name;
            const functionArgs = JSON.parse(toolCall.function.arguments);

            if (functionName === 'get_crypto_price') {
                console.log(`Executing: ${functionName}(${JSON.stringify(functionArgs)})`);

                const functionResponse = await getCryptoPrice(
                    functionArgs.coin_id,
                    functionArgs.currency
                );
                console.log(`Result: ${functionResponse}`);

                messages.push({
                    tool_call_id: toolCall.id,
                    role: 'tool',
                    content: functionResponse,
                });
            }
        }

        console.log('\nSending follow-up request for final answer...');
        const finalResponse = await client.chat.completions.create({
            model: 'gpt-3.5-turbo',
            messages: messages,
        });

        console.log('\nFinal Answer:');
        console.log(finalResponse.choices[0].message.content);
    }
}

main();

File: patterns/tool-use-external-api/openai-like/python/crypto_price.py
import os
import json
import requests
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv()

def get_crypto_price(coin_id, currency="usd"):
    """
    Fetches the current price of a cryptocurrency from CoinGecko.
    """
    url = f"https://api.coingecko.com/api/v3/simple/price?ids={coin_id}&vs_currencies={currency}"
    try:
        response = requests.get(url)
        data = response.json()
        price = data.get(coin_id, {}).get(currency)
        return json.dumps({"coin": coin_id, "price": price, "currency": currency})
    except Exception as e:
        return json.dumps({"error": str(e)})

def main():
    api_key = os.environ.get("OPENAI_API_KEY")
    client = OpenAI(api_key=api_key)

    tools = [
        {
            "type": "function",
            "function": {
                "name": "get_crypto_price",
                "description": "Get the current price of a cryptocurrency",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "coin_id": {
                            "type": "string",
                            "description": "The ID of the coin on CoinGecko (e.g., bitcoin, ethereum, solana)",
                        },
                        "currency": {
                            "type": "string",
                            "description": "The target currency (e.g., usd, eur)",
                        },
                    },
                    "required": ["coin_id"],
                },
            },
        }
    ]

    messages = [{"role": "user", "content": "What is the price of Bitcoin and Solana right now?"}]

    print("User Question:", messages[0]["content"])
    print("Sending request to OpenAI with tools...")

    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=messages,
        tools=tools,
        tool_choice="auto",
    )

    response_message = response.choices[0].message
    tool_calls = response_message.tool_calls

    if tool_calls:
        print(f"Model decided to call {len(tool_calls)} tool(s).")
        messages.append(response_message)

        for tool_call in tool_calls:
            function_name = tool_call.function.name
            function_args = json.loads(tool_call.function.arguments)
            
            if function_name == "get_crypto_price":
                print(f"Executing: {function_name}({function_args})")
                function_response = get_crypto_price(
                    coin_id=function_args.get("coin_id"),
                    currency=function_args.get("currency", "usd")
                )
                print(f"Result: {function_response}")
                
                messages.append(
                    {
                        "tool_call_id": tool_call.id,
                        "role": "tool",
                        "name": function_name,
                        "content": function_response,
                    }
                )

        print("\nSending follow-up request for final answer...")
        final_response = client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=messages,
        )
        print("\nFinal Answer:")
        print(final_response.choices[0].message.content)

if __name__ == "__main__":
    main()
